nav2_behavior_tree/src/behavior_tree_engine.cpp

```diff
@@ -64,13 +63,7 @@ BehaviorTreeEngine::run(
 
       onLoop();
 
-      if (!loopRate.sleep()) {
-        RCLCPP_DEBUG_THROTTLE(
-          rclcpp::get_logger("BehaviorTreeEngine"),
-          *clock_, 1000,
-          "Behavior Tree tick rate %0.2f was exceeded!",
-          1.0 / (loopRate.period().count() * 1.0e-9));
-      }
```

Comment:
We are losing the feature of knowing if we are not reaching the defined tick rate though :/
On the other hand,  with some modification in BT.cpp, we could get from `Tree::sleep`, whether we interrupted the sleep.

Reply:
software implementation-wise, I'm approving this (obviously :wink: )

It does hit me strange yes that we're not trying to smooth out rates. What's nice about rclcpp::Rate is that it'll only sleep enough to maintain a loop rate. This'll sleep for an amount of time regardless of the time it took since the last loop. I think that's less a debugging as much as functionally important. If you say run at 100hz, you don't want to find it running at 92hz or something just due to the fixed sleep being `0.01s` not considering the BT's execution duration itself. The **logging** of it I agree is mostly for debugging, but I figure if we're going to do this, may as well do that (but its not mission critical)

So to get the best behavior, it seems like we should have a `BT::LoopRate` object perhaps that mimic's ROS' version (we can leave sim time out, just wall time) so that it tracks the rate and calls `sleep()` on the `timeout - loop_time` to maintain a fixed loop rate, less interrupted. 

Since that's pretty standalone, that could go back into BT.CPP easily I think, but we could stage it in Nav2 to get it out sooner than later. Then, open a PR with BT.CPP for its review cycle / merge / release that we could migrate to once out in binaries.